lib/auth/accountrecovery.go

```diff
@@ -33,13 +37,113 @@ import (
 const (
 	numOfRecoveryCodes     = 3
 	numWordsInRecoveryCode = 8
+
+	// accountLockedMsg is the reason used to update a user's status locked message.
+	accountLockedMsg = "user has exceeded maximum failed account recovery attempts"
 )
 
 // fakeRecoveryCodeHash is bcrypt hash for "fake-barbaz x 8".
 // This is a fake hash used to mitigate timing attacks against invalid usernames or if user does
 // exist but does not have recovery codes.
 var fakeRecoveryCodeHash = []byte(`$2a$10$c2.h4pF9AA25lbrWo6U0D.ZmnYpFDaNzN3weNNYNC3jAkYEX9kpzu`)
 
+// ErrMaxFailedRecoveryAttempts is a user friendly error message to notify user that recovery attempt
+// has been temporarily locked and an email has been sent.
+var ErrMaxFailedRecoveryAttempts = trace.AccessDenied("too many incorrect attempts, please check your email and try again later")
+
+// StartAccountRecovery implements AuthService.StartAccountRecovery.
+func (s *Server) StartAccountRecovery(ctx context.Context, req *proto.StartAccountRecoveryRequest) (types.UserToken, error) {
```

Comment:
Let's make sure that anything that leaks information about the user or the system returns an AccessDenied error. Feel free to provide details in the logs, but I also suggest we return the same error string in all of them. We don't want this to be used to guess which users exist or callers being able to tell how far they can progress in the method through the error strings.

Please cover it with testing (a single table-driven test if possible).

Reply:
I wasn't too sure about the same error string for all errors. I went back and did the following:

1.  for system errors, I returned a generic error that just says `unable to start account recovery, please try again or contact your system administrator`, while logging out real error
1. for user errors like username not being in valid email format, i also returned the same generic msg as above
1. for user not found and invalid recovery codes, i returned a `invalid username or recovery code` error
2. When user maxed out their attempts, returned `too many incorrect attempts, please try again later`

Not sure I did it right, but i tested for these specific error messages for # 2-4 (i didn't test for system errors)

Reply:
Sounds reasonable. Let's get a 2nd opinion from @awly just in case - Andrew, WDYT?

Reply:
confirmed with andrew on slack that its good to go